hw/ip/otbn/dv/rig/rig/gens/loop.py

```diff
@@ -122,8 +122,10 @@ def _pick_loop_iterations(self,
             # We will start '1 + lo' iterations before the warp (and finish
             # 'lo' of them). We'll then finish 'max_iters - lo' iterations
             # after the warp. Both '1 + lo' and 'max_iters - lo' must be
-            # positive.
-            lo = random.randint(0, max_iters - 1)
+            # positive. `lo` must be at least 1 as if we warp to 0 remaining
+            # iterations right at the end of a loop the prefetch stage will have
+            # fetched the wrong instruction and a fatal error will be raised.
+            lo = random.randint(1, max_iters - 1)
```

Comment:
I'm not completely sure this does what you intended. I think the idea in the original code was that (counting from zero), we warp from iteration `lo` to iteration `hi`. So if `actual_iters` was 1000, `lo` was 1 and `hi` was 998 then we'd execute the loop 3 times (iteration 0; iteration 1, turning into 998; iteration 999).

If you're trying to make sure we don't jump to the final iteration, I think you need to ensure that `hi < actual_iters - 1`. Another way of looking at it is that we run `max_iters` iterations in total and need to do at least 1 full iteration after the jump, meaning that `lo` must be in the range `[0, max_iters - 2]`. At the top end, that would have us warping from `max_iters - 2` to `actual_iters - (max_iters - (max_iters - 2)) = actual_iters - 2`. We'd finish that iteration and then have the whole of iteration `actual_iters - 1` still to do.

Does that sound right to you?

Reply:
Yes, getting confused between this and the RTL iteration count that starts high and goes low.

Reply:
You and me both! I remember taking ages to get this right :-)